sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -701,97 +703,242 @@ public boolean isSnapshot() {
+
+                    return this.downloadWithResponse(new BlobRange(0, 0L), downloadRetryOptions, requestConditions,
+                        rangeGetContentMd5, context)
+                        .subscribeOn(Schedulers.elastic())
+                        .flatMap(response -> {
+                            /*
+                            Ensure the blob is still 0 length by checking our download was the full length.
+                            (200 is for full blob; 206 is partial).
+                             */
+                            if (response.getStatusCode() != 200) {
+                                Mono.error(new IllegalStateException("Blob was modified mid download. It was "
+                                    + "originally 0 bytes and is now larger."));
+                            }
+                            return Mono.zip(Mono.just(0L), Mono.just(requestConditions), Mono.just(response));
+                        });
+                }
+                return Mono.error(blobStorageException);
+            });
     }
 
-    private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath) {
-        try {
-            return AsynchronousFileChannel.open(Paths.get(filePath), StandardOpenOption.READ, StandardOpenOption.WRITE,
-                StandardOpenOption.CREATE_NEW);
-        } catch (IOException e) {
-            throw logger.logExceptionAsError(new UncheckedIOException(e));
-        }
+    private static BlobRequestConditions setEtag(BlobRequestConditions accessConditions, String etag) {
+        //We don't want to modify the user's object, so we'll create a duplicate and set the retrieved etag.
+        return new BlobRequestConditions()
+            .setIfModifiedSince(
+                accessConditions.getIfModifiedSince())
+            .setIfUnmodifiedSince(
+                accessConditions.getIfModifiedSince())
+            .setIfMatch(etag)
+            .setIfNoneMatch(
+                accessConditions.getIfNoneMatch())
+            .setLeaseId(accessConditions.getLeaseId());
+    }
+
+    private static Mono<Void> writeBodyToFile(BlobDownloadAsyncResponse response,
+        AsynchronousFileChannel file, long chunkNum, ParallelTransferOptions optionsReal, Lock progressLock,
+        AtomicLong totalProgress) {
+
+        // Extract the body.
+        Flux<ByteBuffer> data = response.getValue();
+
+        // Report progress as necessary.
+        data = ProgressReporter.addParallelProgressReporting(data,
+            optionsReal.getProgressReceiver(), progressLock, totalProgress);
+
+        // Write to the file.
+        return FluxUtil.writeFile(data, file,
+            chunkNum * optionsReal.getBlockSize());
+    }
+
+    private static Response<BlobProperties> buildBlobPropertiesResponse(BlobDownloadAsyncResponse response) {
+        BlobProperties properties = new BlobProperties(null, response.getDeserializedHeaders().getLastModified(),
```

Comment:
nit: this could be slightly less verbose by making this method return just BlobProperties and only accept the headers as an argument. and wherever you call it you can essentially just say  new SimpleResponse<>(response, buildBlobProperties(response.getDeserializedHeaders()));

Reply:
I only ever call this once. It's just to hide this ugly constructor block. I think it's more concise constructing the response in the method cus then I can hide the response construction, too.